src/impy/kinematics.py

```diff
@@ -184,9 +189,47 @@ def _get_maximum_AZ(self):
         max_ind = a_val.index(max(a_val))
         return self.component_A[max_ind], self.component_Z[max_ind]
 
+    def _get_sample(self, nevt):
+        """Get counts for each realization from self.component_fractions distribution
+        for 'nevt' events"""
+        diff_dist = np.zeros((len(self.component_fractions),), dtype=np.int64)
+        for i in range(nevt):
+            ic = self.rng.choice(self.ncomponents, 1, p=self.component_fractions)[0]
+            diff_dist[ic] += 1
+        self._diff_dist = np.append(diff_dist, 0)
+        self._is_ordered = True
+
+    def _get_next_event_index(self):
+        diff_dist = self._diff_dist
+        j = diff_dist[-1]
+        n = len(diff_dist) - 1
+        res = None
+        if j < n:
+            while j < n and diff_dist[j] == 0:
+                j += 1
+            if j < n:
+                res = j
+            if diff_dist[j] > 0:
+                diff_dist[j] -= 1
+
+        diff_dist[-1] = j
+        return res
```

Comment:
I am not sure what you do here, but this logic should be in MCRun if at all.

Reply:
Now the logic of returning index is in `_gen_target_ind`. Python generator function is used to save a state.